src/libcollections/string.rs

```diff
@@ -1260,6 +1260,36 @@ impl String {
         self.len() == 0
     }
 
+    /// Divide one string into two at an index.
+    ///
+    /// The argument, `mid`, should be a byte offset from the start of the string. It must also
+    /// be on the boundary of a UTF-8 code point.
+    ///
+    /// The two strings returned go from the start of the string to `mid`, and from `mid` to the end
+    /// of the string.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `mid` is not on a `UTF-8` code point boundary, or if it is beyond the last
+    /// code point of the string.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let mut hello = String::from("Hello, World!");
+    /// let world = hello.split_off(7);
+    /// assert_eq!(s, "Hello, ");
+    /// assert_eq!(s2, "World!");
+    /// ```
+    #[inline]
+    #[unstable(feature = "string_split_off", issue = "0")]
+    pub fn split_off(&mut self, mid: usize) -> String {
+        assert!(self.is_char_boundary(mid));
+        assert!(mid <= self.len());
```

Comment:
You also probably want to use `slice_error_fail` rather than plain assertion, because it will result in less code in callers and perhaps a bit better message (even if you specify `mid` for both `start` and `end` arguments of that function).